src/infra/core/src/compact_service_impl.rs

```diff
@@ -0,0 +1,289 @@
+            .await?;
+
+        let add_data_params = AddDataParams {
+            prev_checkpoint: None,
+            prev_offset: last_dataslice_info.prev_offset,
+            new_offset_interval: last_dataslice_info.new_offset_interval.clone(),
+            new_watermark: None,
+            new_source_state: None,
+        };
+
+        dataset
+            .commit_add_data(
+                add_data_params,
+                Some(OwnedFile::new(new_file_path)),
+                None,
+                CommitOpts {
+                    block_ref: &BlockRef::Head,
+                    system_time: Some(Utc::now()),
+                    prev_block_hash: Some(Some(block_hash_to_reset)),
+                    check_object_refs: false,
+                },
+            )
+            .await?;
+
+        Ok(())
+    }
+
+    fn remove_old_files(
+        &self,
+        chain_files_info: &ChainFilesInfo,
+        listener: &Arc<dyn CompactionListener>,
+    ) -> Result<(), CompactError> {
+        listener.begin_phase(CompactionPhase::CleanDatasliceFiles);
+        for file_url in &chain_files_info.data_slice_file_urls {
+            fs::remove_file(file_url.replace("file://", "")).int_err()?;
+        }
+
+        listener.begin_phase(CompactionPhase::CleanBlockFiles);
+        let block_hash_to_reset_string = chain_files_info
+            .block_hash_to_reset
+            .as_multibase()
+            .to_string();
+        for block_file_url in &chain_files_info.block_file_urls {
+            if block_file_url.contains(&block_hash_to_reset_string) {
+                break;
+            }
+            fs::remove_file(block_file_url.replace("file://", "")).int_err()?;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait::async_trait]
+impl CompactService for CompactServiceImpl {
+    #[tracing::instrument(level = "info", skip_all)]
+    async fn compact_dataset(
+        &self,
+        dataset_handle: &DatasetHandle,
+        dataset_dir_path: &Path,
+        multi_listener: Option<Arc<dyn CompactionMultiListener>>,
+    ) -> Result<(), CompactError> {
+        let listener = multi_listener
+            .and_then(|l| l.begin_compact(dataset_handle))
+            .unwrap_or(Arc::new(NullCompactionListener {}));
+        self.dataset_authorizer
+            .check_action_allowed(dataset_handle, domain::auth::DatasetAction::Read)
+            .await?;
+
+        let dataset = self
+            .dataset_repo
+            .get_dataset(&dataset_handle.as_local_ref())
+            .await?;
+
+        let dataset_kind = dataset
+            .get_summary(GetSummaryOpts::default())
+            .await
+            .int_err()?
+            .kind;
+
+        if dataset_kind != DatasetKind::Root {
+            return Err(CompactError::InvalidDatasetKind(InvalidDatasetKindError {
+                dataset_name: dataset_handle.alias.dataset_name.clone(),
+            }));
+        }
+
+        let compact_dir_path = self.create_compact_dir(dataset_dir_path)?;
```

Comment:
Where would I create this in S3 on deployment?

Reply:
I wonder if we could do compaction *within* the same dataset...

Compaction is essentially committing blocks with different data slices. If we support committing a block without modifying the HEAD - we could start essentially creating a branch and then switch `HEAD` to point to that branch as the very last step.

MetadataChain already supports appending blocks without modifying the `HEAD` reference, so we might just need to allow this in `Dataset::commit_*` helper functions too.

This way:
- we don't need any temporary directories
- we preserve original dataset intact until the very last point of updating the HEAD
- if compaction fails / is interrupted - we just have a bit of garbage left for GC to take care of

Reply:
Great point, done